cron/pubsub/subscriber_gcs.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2021 Security Scorecard Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pubsub
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	pubsub "cloud.google.com/go/pubsub/apiv1"
+	pubsubpb "google.golang.org/genproto/googleapis/pubsub/v1"
+
+	"github.com/ossf/scorecard/cron/data"
+)
+
+const (
+	maxMessagesToPull         = 1
+	ackDeadlineExtensionInSec = 600
+	gracePeriodInSec          = 60
+	gcpPubsubPrefix           = "gcppubsub://"
+)
+
+type gcsSubscriber struct {
+	ctx             context.Context
+	done            chan bool
+	client          *pubsub.SubscriberClient
+	pullRequest     *pubsubpb.PullRequest
+	subscriptionURL string
+	recvdAckIDs     []string
+	delay           time.Duration
+}
+
+func createGCSSubscriber(ctx context.Context, subscriptionURL string) (*gcsSubscriber, error) {
+	client, err := pubsub.NewSubscriberClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error during NewSubscriberClient: %w", err)
+	}
+
+	validSubscriptionURL := strings.TrimPrefix(subscriptionURL, gcpPubsubPrefix)
+	ret := &gcsSubscriber{
+		ctx:             ctx,
+		client:          client,
+		subscriptionURL: validSubscriptionURL,
+		pullRequest: &pubsubpb.PullRequest{
+			Subscription: validSubscriptionURL,
+			MaxMessages:  maxMessagesToPull,
+		},
+	}
+	return ret, nil
+}
+
+func (subscriber *gcsSubscriber) extendAckDeadline() {
+	for {
+		select {
+		case <-subscriber.ctx.Done():
+			return
+		case <-subscriber.done:
+			return
+		case <-time.After(subscriber.delay):
+			ackDeadline := ackDeadlineExtensionInSec * time.Second
+			err := subscriber.client.ModifyAckDeadline(subscriber.ctx, &pubsubpb.ModifyAckDeadlineRequest{
+				Subscription:       subscriber.subscriptionURL,
+				AckIds:             subscriber.recvdAckIDs,
+				AckDeadlineSeconds: int32(ackDeadline.Seconds()),
+			})
+			if err != nil {
+				log.Fatal(err)
+			}
+			subscriber.delay = ackDeadline - gracePeriodInSec*time.Second
```

Comment:
Any reason we need to store/recompute this all the time rather than storing them as consts?

Reply:
I want to have it in such a way that the first time we call `extendDeadline()`, `delay = 0`. After which it is a const.